aten/src/ATen/native/sparse/SparseMatMul.cpp

```diff
@@ -0,0 +1,375 @@
+    should be compatible.
+
+    This is an implementation of the SMMP algorithm:
+     "Sparse Matrix Multiplication Package (SMMP)"
+
+       Randolph E. Bank and Craig C. Douglas
+
+     http://citeseer.ist.psu.edu/445062.html
+     http://www.mgnet.org/~douglas/ccd-codes.html
+  */
+  std::vector<int64_t> mask(n_col, -1);
+  const auto long_max = std::numeric_limits<int64_t>::max();
+
+  int64_t nnz = 0;
+  for (int64_t i = 0; i < n_row; i++) {
+    int64_t row_nnz = 0;
+
+    for (int64_t jj = Ap[i]; jj < Ap[i + 1]; jj++) {
+      int64_t j = Aj[jj];
+      for (int64_t kk = Bp[j]; kk < Bp[j + 1]; kk++) {
+        int64_t k = Bj[kk];
+        if (mask[k] != i) {
+          mask[k] = i;
+          row_nnz++;
+        }
+      }
+    }
+    int64_t next_nnz = nnz + row_nnz;
+    TORCH_CHECK(row_nnz <= long_max - nnz, "nnz of the output is too large");
+    nnz = next_nnz;
+  }
+  return nnz;
+}
+
+template<class scalar_t>
+void _csr_matmult(
+    const int64_t n_row,
+    const int64_t n_col,
+    const int64_t Ap[],
+    const int64_t Aj[],
+    const scalar_t Ax[],
+    const int64_t Bp[],
+    const int64_t Bj[],
+    const scalar_t Bx[],
+    int64_t Cp[],
+    int64_t Cj[],
+    scalar_t Cx[]) {
+  /* 
+    Compute CSR entries for matrix C = A*B.
+
+    The matrices `A` and 'B' should be in proper CSR structure, and their dimensions
+    should be compatible.
+
+    Inputs:
+      `n_row`         - number of row in A
+      `n_col`         - number of columns in B
+      `Ap[n_row+1]`   - row pointer
+      `Aj[nnz(A)]`    - column indices
+      `Ax[nnz(A)]     - nonzeros
+      `Bp[?]`         - row pointer
+      `Bj[nnz(B)]`    - column indices
+      `Bx[nnz(B)]`    - nonzeros
+    Outputs:
+      `Cp[n_row+1]` - row pointer
+      `Cj[nnz(C)]`  - column indices
+      `Cx[nnz(C)]`  - nonzeros
+
+    Note:
+      Output arrays Cp, Cj, and Cx must be preallocated
+
+    This is an implementation of the SMMP algorithm:
+     "Sparse Matrix Multiplication Package (SMMP)"
+
+       Randolph E. Bank and Craig C. Douglas
+
+     http://citeseer.ist.psu.edu/445062.html
+     http://www.mgnet.org/~douglas/ccd-codes.html
+
+  */
+  std::vector<int64_t> next(n_col, -1);
+  std::vector<scalar_t> sums(n_col, 0);
+
+  int64_t nnz = 0;
+
+  Cp[0] = 0;
+
+  for (int64_t i = 0; i < n_row; i++) {
+    int64_t head = -2;
+    int64_t length = 0;
+
+    int64_t jj_start = Ap[i];
+    int64_t jj_end = Ap[i + 1];
+    for (int64_t jj = jj_start; jj < jj_end; jj++) {
+      int64_t j = Aj[jj];
+      scalar_t v = Ax[jj];
+
+      int64_t kk_start = Bp[j];
+      int64_t kk_end = Bp[j + 1];
+      for (int64_t kk = kk_start; kk < kk_end; kk++) {
+        int64_t k = Bj[kk];
+
+        sums[k] += v * Bx[kk];
+
+        if (next[k] == -1) {
+          next[k] = head;
+          head = k;
+          length++;
+        }
+      }
+    }
+
+    for (int64_t jj = 0; jj < length; jj++) {
+      if (sums[head] != 0) {
```

Comment:
Thanks for notice this @pearu. This condition was used to skip zeros in the output matrix when it is possible. I get your point when the condition is false then `maxnnz != nnz`.  There was an error here as `nnz` have to be updated after calling  `__csr_matmult_maxnnz`.  Moreover,  I can see that this can change also the sparsity structure, so to maintain consistency in code logic and avoid handling edge cases I will remove this condition.